tunnel-obfuscation/src/shadowsocks.rs

```diff
@@ -0,0 +1,258 @@
+
+    Ok(())
+}
+
+async fn create_shadowsocks_client(
+    shadowsocks_endpoint: SocketAddr,
+    #[cfg(target_os = "linux")] fwmark: Option<u32>,
+) -> std::result::Result<ProxySocket, ProxySocketError> {
+    let ss_context = Context::new_shared(ServerType::Local);
+    let ss_config: ServerConfig = ServerConfig::new(
+        shadowsocks_endpoint,
+        SHADOWSOCKS_PASSWORD,
+        SHADOWSOCKS_CIPHER,
+    );
+    let connect_opts = ConnectOpts {
+        #[cfg(target_os = "linux")]
+        fwmark,
+        ..Default::default()
+    };
+    ProxySocket::connect_with_opts(ss_context, &ss_config, &connect_opts).await
+}
+
+async fn create_local_udp_socket(ipv4: bool) -> Result<(UdpSocket, SocketAddr)> {
+    let random_bind_addr = if ipv4 {
+        SocketAddr::new("127.0.0.1".parse().unwrap(), 0)
+    } else {
+        SocketAddr::new("::1".parse().unwrap(), 0)
+    };
+    let local_udp_socket = UdpSocket::bind(random_bind_addr)
+        .await
+        .map_err(Error::BindUdp)?;
+    let udp_client_addr = local_udp_socket
+        .local_addr()
+        .map_err(Error::GetUdpLocalAddress)?;
+
+    Ok((local_udp_socket, udp_client_addr))
+}
+
+/// Wait for a client to connect to `udp_listener` and connect the socket to that address
+async fn wait_for_local_udp_client(udp_listener: &UdpSocket) -> io::Result<()> {
+    let mut client_rx_buffer = vec![0u8; u16::MAX as usize];
+
+    log::trace!("Waiting for UDP socket client");
+    let (_bytes_n, client_addr) = udp_listener.peek_from(&mut client_rx_buffer).await?;
+
+    log::trace!("UDP connection from {client_addr}");
+    udp_listener.connect(client_addr).await
+}
+
+async fn handle_outgoing(
+    ss_write: Arc<ProxySocket>,
+    local_udp_read: Arc<UdpSocket>,
+    wg_addr: Address,
+) {
+    let mut rx_buffer = vec![0u8; u16::MAX as usize];
+
+    loop {
+        let read_n = match local_udp_read.recv(&mut rx_buffer).await {
+            Ok(read_n) => read_n,
+            Err(error) => {
+                log::error!("Failed to read from local UDP socket: {error}");
+                break;
+            }
+        };
+
+        if let Err(error) = ss_write.send(&wg_addr, &rx_buffer[0..read_n]).await {
+            log::error!("Failed to write to Shadowsocks client: {error}");
+            if is_fatal_socket_error(&error) {
+                break;
+            }
+        }
+    }
+}
+
+async fn handle_incoming(
+    ss_read: Arc<ProxySocket>,
+    local_udp_write: Arc<UdpSocket>,
+    wg_addr: Address,
+) {
+    let mut rx_buffer = vec![0u8; u16::MAX as usize];
+
+    loop {
+        let (read_n, addr, _ctrl) = match ss_read.recv(&mut rx_buffer).await {
+            Ok(read_n) => read_n,
+            Err(error) => {
+                log::error!("Failed to read from Shadowsocks client: {error}");
+                break;
+            }
+        };
+
+        if addr != wg_addr {
+            log::trace!("Ignoring packet from unexpected source: {addr}");
+            continue;
+        }
```

Comment:
Silly question maybe, but how come the address we recieve from will be of the wireguard endpoint and not the shadowsocks address? Should it be possible to receive from another address? If not, maybe the log level should be `warn`.

Reply:
This is the address in the shadowsocks layer, not the IP header. Technically, nothing stops the shadowsocks server from trying to forward packets from some other source. It shouldn't ever happen, though.

I set it to trace because if it did occur, we'd log this for every single packet received. :shrug: It's probably better not to trust servers not to flood our logs.